### ISSUE TITLE
HR fails when removing a property with an initializer from a partial class

### ISSUE THREAD

**Issue by dr1rrb** [dotnet/roslyn#79320]
**Version Used**: net 9.0 / Microsoft.CodeAnalysis 4.14.0

**Steps to Reproduce**:
1. Start the repro project [RoslynOutOfRange.zip](https://github.com/user-attachments/files/21150407/RoslynOutOfRange.zip) with debugger attached
2. In the file `MainPage.g.cs` comment the line 11 `private object _component_0_Holder { get; set; } = new();`
3. Invoke HR
4. Observe error
`error ENC1002: Cannot apply changes -- unexpected error: 'Index was outside the bounds of the array.'`

<img width="902" height="283" alt="Image" src="https://github.com/user-attachments/assets/7eb29ccc-e648-47bd-b3a2-149e9f026992" />

**Diagnostic Id**: ENC1002

If this is a report about a bug in an analyzer, please include the diagnostic ID and message if possible (e.g. `"IDE0030: Use coalesce expression"`).

**Expected Behavior**:
Able to apply changes using hot-reload.

**Actual Behavior**:
`error ENC1002: Cannot apply changes -- unexpected error: 'Index was outside the bounds of the array.'`

**Comment by dr1rrb** [comment]
For reference, the complete call-stack (includes our internal testing engine):

```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Collections.Immutable.ImmutableArray`1.get_Item(Int32 index)
   at Microsoft.CodeAnalysis.EditAndContinue.EditSession.<>c__DisplayClass26_0.<MergePartialEdits>b__5(SyntaxNode node)
   at Microsoft.CodeAnalysis.Emit.EncVariableSlotAllocator.TryGetPreviousLambdaSyntaxOffset(SyntaxNode lambdaOrLambdaBodySyntax, Boolean isLambdaBody, Int32& previousSyntaxOffset)
   at Microsoft.CodeAnalysis.Emit.EncVariableSlotAllocator.TryGetPreviousLambda(SyntaxNode lambdaOrLambdaBodySyntax, Boolean isLambdaBody, Int32 closureOrdinal, ImmutableArray`1 structClosureIds, DebugId& lambdaId, Nullable`1& runtimeRudeEdit)
   at Microsoft.CodeAnalysis.CSharp.ClosureConversion.GetLambdaId(SyntaxNode syntax, ClosureKind closureKind, Int32 closureOrdinal, ImmutableArray`1 structClosureIds, Nullable`1 closureRudeEdit)
   at Microsoft.CodeAnalysis.CSharp.ClosureConversion.<SynthesizeClosureMethods>b__37_0(Scope scope, NestedFunction nestedFunction)
   at Microsoft.CodeAnalysis.CSharp.ClosureConversion.Analysis.VisitNestedFunctions(Scope scope, Action`2 action)
   at Microsoft.CodeAnalysis.CSharp.ClosureConversion.SynthesizeClosureMethods()
   at Microsoft.CodeAnalysis.CSharp.ClosureConversion.Rewrite(BoundStatement loweredBody, NamedTypeSymbol thisType, ParameterSymbol thisParameter, MethodSymbol method, Int32 methodOrdinal, MethodSymbol substitutedSourceMethod, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 lambdaRuntimeRudeEditsBuilder, ArrayBuilder`1 closureDebugInfoBuilder, VariableSlotAllocator slotAllocator, TypeCompilationState compilationState, BindingDiagnosticBag diagnostics, HashSet`1 assignLocals)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, SourceExtensionImplementationMethodSymbol extensionImplementationMethod, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, MethodInstrumentation instrumentation, DebugDocumentProvider debugDocumentProvider, ImmutableArray`1& codeCoverageSpans, BindingDiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 lambdaRuntimeRudeEditsBuilder, ArrayBuilder`1 closureDebugInfoBuilder, ArrayBuilder`1 stateMachineStateDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass25_0.<CompileNamedTypeAsync>b__0()
   at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers()
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean emittingPdb, Boolean hasDeclarationErrors, Boolean emitMethodBodies, BindingDiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Compile(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Emit.EmitHelpers.EmitDifference(CSharpCompilation compilation, EmitBaseline baseline, IEnumerable`1 edits, Func`2 isAddedSymbol, Stream metadataStream, Stream ilStream, Stream pdbStream, CompilationTestData testData, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.EmitDifference(EmitBaseline baseline, IEnumerable`1 edits, Func`2 isAddedSymbol, Stream metadataStream, Stream ilStream, Stream pdbStream, CompilationTestData testData, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.EmitDifference(EmitBaseline baseline, IEnumerable`1 edits, Func`2 isAddedSymbol, Stream metadataStream, Stream ilStream, Stream pdbStream, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.EditAndContinue.EditSession.EmitSolutionUpdateAsync(Solution solution, ActiveStatementSpanProvider solutionActiveStatementSpanProvider, UpdateId updateId, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.EditAndContinue.DebuggingSession.EmitSolutionUpdateAsync(Solution solution, IImmutableSet`1 runningProjects, ActiveStatementSpanProvider activeStatementSpanProvider, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.ExternalAccess.Watch.Api.WatchHotReloadService.GetUpdatesAsync(Solution solution, IImmutableSet`1 runningProjects, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.ExternalAccess.Watch.Api.WatchHotReloadService.EmitSolutionUpdateAsync(Solution solution, CancellationToken cancellationToken)
   at Uno.UI.RemoteControl.Host.HotReload.MetadataUpdates.WatchHotReloadService.<>c__DisplayClass5_0.<<-ctor>b__0>d.MoveNext() in C:\Src\GitHub\unoplatform\uno.4\src\Uno.UI.RemoteControl.Server.Processors\HotReload\MetadataUpdates\WatchHotReloadService.cs:line 71
--- End of stack trace from previous location ---
   at Uno.UI.RemoteControl.Host.HotReload.MetadataUpdates.WatchHotReloadService.EmitSolutionUpdateAsync(Solution solution, CancellationToken cancellationToken) in C:\Src\GitHub\unoplatform\uno.4\src\Uno.UI.RemoteControl.Server.Processors\HotReload\MetadataUpdates\WatchHotReloadService.cs:line 121
   at Uno.UI.SourceGenerators.MetadataUpdates.HotReloadWorkspace.Update() in C:\Src\GitHub\unoplatform\uno.4\src\SourceGenerators\Uno.UI.SourceGenerators.Tests\MetadataUpdates\HotReloadWorkspaceProvider.cs:line 418
   at Uno.UI.SourceGenerators.Tests.MetadataUpdateTests.Given_HotReloadService.ApplyScenario(Project[] projects, Boolean isDebugCompilation, Boolean isMono, Boolean useXamlReaderReload, String name, CancellationToken ct) in C:\Src\GitHub\unoplatform\uno.4\src\SourceGenerators\Uno.UI.SourceGenerators.Tests\MetadataUpdateTests\Given_HotReloadService.cs:line 220
   at Uno.UI.SourceGenerators.Tests.MetadataUpdateTests.Given_HotReloadService.HR(String name, Scenario scenario, Project[] projects) in C:\Src\GitHub\unoplatform\uno.4\src\SourceGenerators\Uno.UI.SourceGenerators.Tests\MetadataUpdateTests\Given_HotReloadService.cs:line 24
```

**Comment by RikkiGibson** [comment]
~~@jcouv I am curious if the current work on extension type name stability will improve this scenario~~
Nevermind, I misread the issue. Extensions are not involved

**Comment by jeromelaban** [comment]
Thanks @tmat!
